token-contracts/OToken/src/main/java/finance/omm/score/tokens/OTokenImpl.java

```diff
@@ -0,0 +1,550 @@
+                    exaMultiply(convertToExa(balance, actualDecimals),
+                            Context.call(BigInteger.class,
+                                    lendingPoolCoreAddress,
+                                    "getNormalizedIncome",
+                                    reserveAddress)),
+                    userIndex);
+            return convertExaToOther(newBalance, actualDecimals.intValue());
+        }
+    }
+
+    protected Map<String, BigInteger> cumulateBalanceInternal(Address user) {
+        BigInteger previousPrincipalBalance = principalBalanceOf(user);
+        BigInteger balanceIncrease = this.balanceOf(user).subtract(previousPrincipalBalance);
+        if (balanceIncrease.compareTo(ZERO) > 0) {
+            this.mint(user, balanceIncrease);
+        }
+
+        Address lendingPoolCoreAddress = this._addresses.get(Contracts.LENDING_POOL_CORE.getKey());
+        Address reserveAddress = this._addresses.get(Contracts.RESERVE.getKey());
+
+        if (lendingPoolCoreAddress == null) {
+            Context.revert(Contracts.LENDING_POOL_CORE.getKey() + " is not configured");
+        }
+
+        if (reserveAddress == null) {
+            Context.revert(Contracts.RESERVE.getKey() + " is not configured");
+        }
+
+        BigInteger userIndex = Context.call(BigInteger.class, lendingPoolCoreAddress, "getNormalizedIncome", reserveAddress);
+        this.userIndexes.set(user, userIndex);
+        return Map.of(
+            "previousPrincipalBalance", previousPrincipalBalance,
+            "principalBalance", previousPrincipalBalance.add(balanceIncrease),
+            "balanceIncrease", balanceIncrease,
+            "index", userIndex);
+    }
+
+    /** This will always include accrued interest as a computed value
+     *
+     */
+    @External(readonly = true)
+    public BigInteger balanceOf(Address _owner) {
+        BigInteger currentPrincipalBalance = this.principalBalanceOf(_owner);
+
+        return calculateCumulatedBalanceInternal(_owner, currentPrincipalBalance);
+    }
+
+    /** This shows the state updated balance and includes the accrued interest upto the most recent computation
+    * initiated by the user transaction
+    */
+    @External(readonly = true)
+    public BigInteger principalBalanceOf(Address _user) {
+        return this.balances.getOrDefault(_user, ZERO);
+    }
+
+    /** The transfer is only allowed if transferring this amount of the underlying collateral doesn't bring the health
+    * factor below 1
+    * */
+    @External(readonly = true)
+    public boolean isTransferAllowed(Address _user, BigInteger _amount) {
+        Address lendingPoolDataProviderAddress = this._addresses.get(Contracts.LENDING_POOL_DATA_PROVIDER.getKey());
+        Address reserveAddress = this._addresses.get(Contracts.RESERVE.getKey());
+
+        if (lendingPoolDataProviderAddress == null) {
+            Context.revert(Contracts.LENDING_POOL_CORE.getKey() + " is not configured");
```

Comment:
LENDING POOL DATA PROVIDER in revert message

Reply:
done